keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/NUTSSampler.java

```diff
@@ -120,288 +114,47 @@ public void step() {
-            1.0,
-            Math.exp(logOfMasterPMinusMomentum - logOfMasterPMinusMomentumBeforeLeapfrog)
-        );
-
-        return new BuiltTree(
-            leapfrog.position,
-            leapfrog.gradient,
-            leapfrog.momentum,
-            leapfrog.position,
-            leapfrog.gradient,
-            leapfrog.momentum,
-            leapfrog.position,
-            leapfrog.gradient,
-            logOfMasterPAfterLeapfrog,
-            sampleAtAcceptedPosition,
-            acceptedLeapfrogCount,
-            shouldContinueFlag,
-            deltaLikelihoodOfLeapfrog,
-            1
-        );
-    }
-
-    private static void acceptOtherPositionWithProbability(double probability,
-                                                           BuiltTree tree,
-                                                           BuiltTree otherTree,
-                                                           KeanuRandom random) {
-        if (withProbability(probability, random)) {
-            tree.acceptedPosition = otherTree.acceptedPosition;
-            tree.gradientAtAcceptedPosition = otherTree.gradientAtAcceptedPosition;
-            tree.logOfMasterPAtAcceptedPosition = otherTree.logOfMasterPAtAcceptedPosition;
-            tree.sampleAtAcceptedPosition = otherTree.sampleAtAcceptedPosition;
+        if (this.adaptEnabled) {
+            stepsize.adaptStepSize(tree, sampleNum);
         }
-    }
 
-    private static boolean withProbability(double probability, KeanuRandom random) {
-        return random.nextDouble() < probability;
+        tree.acceptPositionAndGradient();
+        sampleNum++;
     }
 
-    private static boolean isNotUTurning(Map<VertexId, DoubleTensor> positionForward,
-                                         Map<VertexId, DoubleTensor> positionBackward,
-                                         Map<VertexId, DoubleTensor> momentumForward,
-                                         Map<VertexId, DoubleTensor> momentumBackward) {
-        double forward = 0.0;
-        double backward = 0.0;
-
-        for (Map.Entry<VertexId, DoubleTensor> forwardPositionForLatent : positionForward.entrySet()) {
-
-            final VertexId latentId = forwardPositionForLatent.getKey();
-            final DoubleTensor forwardMinusBackward = forwardPositionForLatent.getValue().minus(
-                positionBackward.get(latentId)
-            );
-
-            forward += forwardMinusBackward.times(momentumForward.get(latentId)).sum();
-            backward += forwardMinusBackward.timesInPlace(momentumBackward.get(latentId)).sum();
-        }
-
-        return (forward >= 0.0) && (backward >= 0.0);
+    private void recordSamplerStatistics() {
+        statistics.store("stepSize", stepsize.getStepsize());
+        statistics.store("logProb", tree.getLogOfMasterPAtAcceptedPosition());
+        statistics.store("treeSize", tree.getTreeSize());
+        statistics.store("meanTreeAccept", stepsize.getAverageAcceptanceProb());
```

Comment:
I'm not a fan of using Strings as a key for these stats, it doesn't seem to buy much in terms of clarity and it's going to be quite costly to have to hash that every time you want to increment stats (yeah, it is optional for us to turn these on, but given they're useful it would be good to have them on by default IMO).

Perhaps instead each sampler should implement their own stats class that just has hard coded methods for each stat and then we have a statistics interface that converts that in to a format suitable for consumption (for instance in to this style of map) once when stats collection is complete?

Reply:
Agreed - changed it to be an ENUM. The sampler defines the set of ENUM's that are passed to the stats class